src/ng/directive/as.js

```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+/**
+ * @ngdoc directive
+ * @name as
+ * @restrict A
+ *
+ * @description
+ * Publishes a component controller or a DOM element into the current scope.
+ *
+ * @element ANY
+ * @param {expression} as {@link guide/expression Expression} to assign the controller.
+ *
+ * This directive allows to access to the controller of components in your
+ * current template or from current controller.
+ *
+ * @example
+ * ### Access to component controller
+ * This example demonstrates basic publishing of the component controller into the scope.
+ * <example name="asDirectiveComponentExample" module="asComponentExample">
+ *   <file name="index.html">
+ *     <toggle as="myToggle"></toggle>
+ *     <button ng-click="myToggle.toggle()">Toggle</button>
+ *     <div ng-show="myToggle.isOpen()">You are using a children component to show it.</div>
+ *   </file>
+ *   <file name="script.js">
+ *     angular.module('asComponentExample', [])
+ *      .component('toggle', {
+ *         controller: function() {
+ *           var opened = false;
+ *           this.isOpen = function() { return opened; };
+ *           this.toggle = function() { opened = !opened; };
+ *         }
+ *       });
+ *   </file>
+ *   <file name="protractor.js" type="protractor">
+ *      it('should publish the toggle into the scope', function() {
+ *        var toggle = element(by.buttonText('Toggle'));
+ *        expect(toggle.evaluate('myToggle.isOpen()')).toEqual(false);
+ *        toggle.click();
+ *        expect(toggle.evaluate('myToggle.isOpen()')).toEqual(true);
+ *      });
+ *   </file>
+ * </example>
+ */
+var componentAsDirective = ['$parse', function($parse) {
+  return {
+    restrict: 'A',
+    compile: function(tElement, tAttrs) {
+      // gets the expected controller name, converts <x-some-thing> into "someThing")
+      var controllerName = tAttrs.$normalize(tElement[0].tagName.toLowerCase())
+        .replace(/[^\w]\w/g, function(a) { return a.slice(1).toUpperCase(); });
+
+      // special case: ngForm is the only known directive that uses an undocumented
+      // feature of .directive able to change the name of the directive
+      if (controllerName === 'ngForm') {
+        controllerName = 'form';
+      }
+
+      // get the setter for the as attribute
+      var setter = $parse(tAttrs.as).assign;
+
+      return function(scope, iElement, iAttrs) {
+        // gets the controller of the current element (see jqLiteController for details)
+        var controller = iElement.data('$' + controllerName + 'Controller');
```

Comment:
Ops!
Tests with jqLite are fine, but jQuery tests are not.
```javascript
        // this is failing with test:jquery, working with test:jqLite
        var controller = jqLiteData(iElement[0], '$' + controllerName + 'Controller');
```

Although this is working:
```javascript
       // working in both cases
       var controller = jqLite(iElement[0]).data('$' + controllerName + 'Controller');
```

After some research in the code I found the following:
(see https://github.com/angular/angular.js/blob/4883e957971f1c551a0f560427bba0224f291449/src/ng/compile.js#L2333-L2342 )

```javascript
  // compile.js 
          var controllerResult = controller();
          if (controllerResult !== controller.instance) {
            // If the controller constructor has a return value, overwrite the instance
            // from setupControllers
            controller.instance = controllerResult;
            $element.data('$' + controllerDirective.name + 'Controller', controllerResult);
            removeControllerBindingWatches && removeControllerBindingWatches();
            removeControllerBindingWatches =
              initializeDirectiveBindings(controllerScope, attrs, controller.instance, bindings, controllerDirective);
          }
```

And the `$element` that it is using is the same `$element` that postLink function seems to be receiving:
( see https://github.com/angular/angular.js/blob/4883e957971f1c551a0f560427bba0224f291449/src/ng/compile.js#L2381-L2391 )

```javascript
        // POSTLINKING
        for (i = postLinkFns.length - 1; i >= 0; i--) {
          linkFn = postLinkFns[i];
          invokeLinkFn(linkFn,
              linkFn.isolateScope ? isolateScope : scope,
              $element,
              attrs,
              linkFn.require && getControllers(linkFn.directiveName, linkFn.require, $element, elementControllers),
              transcludeFn
          );
        }
```

After doing some checking, `jqLite()` is a wrapper for `jQuery()` when jQuery is present (basically I checked the versions and data function):

```javascript
   // if they were different
   var jqElement = jqLite(iElement[0]);
   expect(jqElement.data).not.toBe(iElement.data);
   expect(jqElement.jquery).not.toBe(iElement.jquery);
   // both fails always
```

My conclussion, as I understand, when jQuery is loaded, I guess because of coding coherence, `compile` uses `jQuery.data()` instead of real `jqLite.data()`. If I'm try to get the data from the `jqLiteExpandoStore` it will fail because it was never user to save this information.

So it seems that it would not be changed.